sql: don't enable "kv tracing" for SHOW TRACE FOR <stmt>

andreimatei opened the pull request:
A "kv trace" is the a set of log messages around SQL<->KV interaction.
The concept is not very well defined at the moment, but it comprises two
categories of messages:
1. Details of KV requests being sent for execution, in particular
pretty-printed scan requests.
2. Per-row messages. In particular, every row that's being returned by
KV for processing by SQL.

These messages are the only ones that are returned as results of
SHOW _KV_ TRACE FOR <stmt>/SESSION.
The situation gets interesting for SHOW TRACE FOR ... (no KV specified).
In this case, before this patch, all the "kv trace" messages were
included in the results. This patch changes that and eliminates the
per-row messages. This is an improvement, per-row messages (as opposed
to per-RPC or higher level) is too verbose for most people that want
traces.

Tactically, a couple of "kv messages" that weren't per row are now
always logged (and hence traced) regardless of whether kv tracing is
enabled or not, and otherwise kv tracing is now no longer enabled for
SHOW TRACE FOR.

This was already the case before this patch, but now this comes to the
foreground: "kv tracing" means one thing externally (as in SHOW KV TRACE
FOR ...), versus internally, in the code, where "kvTracedEnabled" and
similarly-named members refer to "per-row" messages. A round of renames
is in order, but that's left for another day.

Benchstat - no difference
https://gist.github.com/andreimatei/c8ed8f850b59a9eec67b2fb190160108

This patch has no influence on the new index-join benchmark:
```
make bench PKG=./pkg/sql BENCHES=BenchmarkIndexJoin TESTFLAGS='-count 10'
make bench PKG=./pkg/sql BENCHES=BenchmarkIndexJoin TESTFLAGS='-test.benchmem -count 10'

[andrei:~/work/src/github.com/cockroachdb/cockroach] sql-kv-tracing+ ± benchstat /tmp/{before,after}
name         old time/op  new time/op  delta
IndexJoin-8  8.54ms ± 4%  8.54ms ± 1%   ~     (p=0.605 n=9+9)

[andrei:~/work/src/github.com/cockroachdb/cockroach] sql-kv-tracing+ ± benchstat /tmp/{before-mem,after-mem}
name         old time/op    new time/op    delta
IndexJoin-8    8.64ms ± 2%    8.59ms ± 6%    ~     (p=0.243 n=9+10)

name         old alloc/op   new alloc/op   delta
IndexJoin-8    3.62MB ± 0%    3.62MB ± 0%  -0.03%  (p=0.000 n=9+9)

name         old allocs/op  new allocs/op  delta
IndexJoin-8     11.6k ± 0%     11.6k ± 0%    ~     (p=0.314 n=10+10)
```

andreimatei committed:
sql: don't enable "kv tracing" for SHOW TRACE FOR <stmt>

A "kv trace" is the a set of log messages around SQL<->KV interaction.
The concept is not very well defined at the moment, but it comprises two
categories of messages:
1. Details of KV requests being sent for execution, in particular
pretty-printed scan requests.
2. Per-row messages. In particular, every row that's being returned by
KV for processing by SQL.

These messages are the only ones that are returned as results of
SHOW _KV_ TRACE FOR <stmt>/SESSION.
The situation gets interesting for SHOW TRACE FOR ... (no KV specified).
In this case, before this patch, all the "kv trace" messages were
included in the results. This patch changes that and eliminates the
per-row messages. This is an improvement, per-row messages (as opposed
to per-RPC or higher level) is too verbose for most people that want
traces.

Tactically, a couple of "kv messages" that weren't per row are now
always logged (and hence traced) regardless of whether kv tracing is
enabled or not, and otherwise kv tracing is now no longer enabled for
SHOW TRACE FOR.

This was already the case before this patch, but now this comes to the
foreground: "kv tracing" means one thing externally (as in SHOW KV TRACE
FOR ...), versus internally, in the code, where "kvTracedEnabled" and
similarly-named members refer to "per-row" messages. A round of renames
is in order, but that's left for another day.

This patch also introduces log.ExpensiveLogEnabled(), a replacement for
log.V() which also takes tracing into consideration.

This patch has no influence on the new index-join benchmark:

make bench PKG=./pkg/sql BENCHES=BenchmarkIndexJoin TESTFLAGS='-count 10'
make bench PKG=./pkg/sql BENCHES=BenchmarkIndexJoin TESTFLAGS='-test.benchmem -count 10'

[andrei:~/work/src/github.com/cockroachdb/cockroach] sql-kv-tracing+ ± benchstat /tmp/{before,after}
name         old time/op  new time/op  delta
IndexJoin-8  8.54ms ± 4%  8.54ms ± 1%   ~     (p=0.605 n=9+9)

[andrei:~/work/src/github.com/cockroachdb/cockroach] sql-kv-tracing+ ± benchstat /tmp/{before-mem,after-mem}
name         old time/op    new time/op    delta
IndexJoin-8    8.64ms ± 2%    8.59ms ± 6%    ~     (p=0.243 n=9+10)

name         old alloc/op   new alloc/op   delta
IndexJoin-8    3.62MB ± 0%    3.62MB ± 0%  -0.03%  (p=0.000 n=9+9)

name         old allocs/op  new allocs/op  delta
IndexJoin-8     11.6k ± 0%     11.6k ± 0%    ~     (p=0.314 n=10+10)

andreimatei committed:
sql: remove unused args

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/19008)
<!-- Reviewable:end -->

andreimatei commented:
I probably want to cherry-pick this in some 1.1.x release.

jordanlewis commented:
Ah I see - this explains why `SHOW KV TRACE` actually currently returns *more* stuff than `SHOW TRACE`, which seems a little unintuitive to me. I like this idea.

I'm surprised that no logic tests change here. I think we should have at least a modicum of test coverage - our traces are actually a very important feature that we can't allow regression in.

---

Reviewed 2 of 2 files at r1.
Review status: 2 of 8 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/sql/session.go, line 1427 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/19008#-KvZE1kRd2tBHp81l8Hi:-KvZE1kSYHwxVYOSELQ-:bfz4vpl) ([raw file](https://github.com/cockroachdb/cockroach/blob/eca98ea2a80df677baa30f90cd5076d8bae76299/pkg/sql/session.go#L1427)):*
> ```Go
> //   verbose messages around the interaction of SQL with KV. Some of the messages
> //   are per-row.
> func (st *SessionTracing) StartTracing(recType tracing.RecordingType, KVTracingEnabled bool) error {
> ```

nit: capitalized argument names seem weird, or at least not "Roman" with respect to our codebase's Rome...

---

*[pkg/sql/trace.go, line 44 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/19008#-KvZFVY_8ozvP4QZxsFF:-KvZFVY_8ozvP4QZxsFG:b2yzult) ([raw file](https://github.com/cockroachdb/cockroach/blob/eca98ea2a80df677baa30f90cd5076d8bae76299/pkg/sql/trace.go#L44)):*
> ```Go
> 	// If set, the trace will also include "KV trace" messages - verbose messages
> 	// around the interaction of SQL with KV. Some of the messages are per-row.
> 	KVTracingEnabled bool
> ```

nit: i don't think this should be capitalized since it's not exported. Or is it?

---

*[pkg/sql/trace.go, line 59 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/19008#-KvZFZMyue5HLNvNiAkj:-KvZFZMyue5HLNvNiAkk:b-t3slf1) ([raw file](https://github.com/cockroachdb/cockroach/blob/eca98ea2a80df677baa30f90cd5076d8bae76299/pkg/sql/trace.go#L59)):*
> ```Go
> //   verbose messages around the interaction of SQL with KV. Some of the
> //   messages are per-row.
> func (p *planner) makeTraceNode(plan planNode, KVTracingEnabled bool) (planNode, error) {
> ```

nit: caps again.

---

*[pkg/sql/sqlbase/kvfetcher.go, line 227 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/19008#-KvZF0ROAeMSnNp1agq7:-KvZF0ROAeMSnNp1agq8:bxwy0u4) ([raw file](https://github.com/cockroachdb/cockroach/blob/eca98ea2a80df677baa30f90cd5076d8bae76299/pkg/sql/sqlbase/kvfetcher.go#L227)):*
> ```Go
> 	}
> 
> 	scanStr := "Scan "
> ```

Hmmm.... not sure if this makes sense. This emits at least 1 Scan event per read operation. Isn't that pretty expensive?

In general it's not obvious to me that we should ever be doing this much unconditional tracing. But I guess that's another fish to fry - we already do a ton of unconditional tracing (including internals like time spent in the command queue). 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19008#-:-KvZFrOk7RXvucKvP6Uo:bpk69s4)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
> Ah I see - this explains why SHOW KV TRACE actually currently returns more stuff than SHOW TRACE, which seems a little unintuitive to me. I like this idea.

Well, no; before this patch, `SHOW TRACE FOR <stmt>` would include all the messages from `SHOW KV TRACE FOR <stmt>`, so the former should always be a superset. This patch makes the two uncomparable: `SHOW KV TRACE FOR <stmt>` will return a restricted set of messages, but include some per-row ones. The per-row ones are excluded from `SHOW TRACE FOR <session>`. Similarly with the recording done for `SET tracing cluster` vs `SET tracing kv`.
If this is not what you currently see, show me.

"KV tracing" is generally a concept that I have mixed feelings on (I was not involved in creating it). I don't have a clear opinion on whether it deserves to exists and, if it does, exactly what semantics it should have. The thing that I'm interested in and that this patch does, however, is for non-kv tracing to not include per-row messages.


> I'm surprised that no logic tests change here. I think we should have at least a modicum of test coverage - our traces are actually a very important feature that we can't allow regression in.

What test did you have in mind exactly? There is test coverage for tracing, kv and regular. However, the tests are never about checking exactly what a trace looks like; that'd be unwieldy. This PR eliminates a few messages from non-kv traces. How did you expect that to have been tested?
I don't see a particular need for extra testing, but if you have a reasonable idea, I'm opened.

---

Review status: 2 of 8 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/sql/session.go, line 1427 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/19008#-KvZE1kRd2tBHp81l8Hi:-Kv_HF9s3eGhNNy17Z0t:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/eca98ea2a80df677baa30f90cd5076d8bae76299/pkg/sql/session.go#L1427)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

nit: capitalized argument names seem weird, or at least not "Roman" with respect to our codebase's Rome...
</blockquote></details>

Done.

---

*[pkg/sql/trace.go, line 44 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/19008#-KvZFVY_8ozvP4QZxsFF:-Kv_GWz_IZ7MoFj7YDKJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/eca98ea2a80df677baa30f90cd5076d8bae76299/pkg/sql/trace.go#L44)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

nit: i don't think this should be capitalized since it's not exported. Or is it?
</blockquote></details>

Done.

---

*[pkg/sql/trace.go, line 59 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/19008#-KvZFZMyue5HLNvNiAkj:-Kv_H95i8H1aInlfeMyH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/eca98ea2a80df677baa30f90cd5076d8bae76299/pkg/sql/trace.go#L59)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

nit: caps again.
</blockquote></details>

Done.

---

*[pkg/sql/sqlbase/kvfetcher.go, line 227 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/19008#-KvZF0ROAeMSnNp1agq7:-Kv_Er6ueeVZHWqhpMEx:b79zxk1) ([raw file](https://github.com/cockroachdb/cockroach/blob/eca98ea2a80df677baa30f90cd5076d8bae76299/pkg/sql/sqlbase/kvfetcher.go#L227)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Hmmm.... not sure if this makes sense. This emits at least 1 Scan event per read operation. Isn't that pretty expensive?

In general it's not obvious to me that we should ever be doing this much unconditional tracing. But I guess that's another fish to fry - we already do a ton of unconditional tracing (including internals like time spent in the command queue). 
</blockquote></details>

The key here is "per read operation" - per batch sent for execution. These are big operations. Generally they're RPCs. Each one of them generates tons of messages already. One more here is not a big deal.

Unconditional logging is not particularly expensive when tracing and the respective logging is turned off - we've optimized this a bunch (when tracing/logging is enabled, that's another story). I believe we ran some tests about it. There must be some overhead for the boxing that the untyped `Eventf()` interface must do... We should test again at some point.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19008#-:-Kv_BcLL8N9Iqiy6PBuT:bmu3wiq)*
<!-- Sent from Reviewable.io -->

knz commented:
Reviewed 2 of 2 files at r1, 6 of 6 files at r3, 6 of 6 files at r4.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/sql/sqlbase/kvfetcher.go, line 227 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/19008#-KvZF0ROAeMSnNp1agq7:-Kva1sHcq30tTLK-az4S:bw106x7) ([raw file](https://github.com/cockroachdb/cockroach/blob/eca98ea2a80df677baa30f90cd5076d8bae76299/pkg/sql/sqlbase/kvfetcher.go#L227)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

The key here is "per read operation" - per batch sent for execution. These are big operations. Generally they're RPCs. Each one of them generates tons of messages already. One more here is not a big deal.

Unconditional logging is not particularly expensive when tracing and the respective logging is turned off - we've optimized this a bunch (when tracing/logging is enabled, that's another story). I believe we ran some tests about it. There must be some overhead for the boxing that the untyped `Eventf()` interface must do... We should test again at some point.

</blockquote></details>

Yes we have ran tests for this and yes it is rather expensive. Sufficiently expensive in fact that Peter had added a separate boolean for this block of code (and related code) before, so it wouldn't run on the regular code path. Check the history of the code. @petermattis what do you think? Right now this code is alloc-heavy (especially the string appends). At the least use a bytes.Buffer.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19008)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 2 of 8 files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/sql/sqlbase/kvfetcher.go, line 227 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/19008#-KvZF0ROAeMSnNp1agq7:-KvbreBFajSmVfKnMc6o:b7o8e7l) ([raw file](https://github.com/cockroachdb/cockroach/blob/eca98ea2a80df677baa30f90cd5076d8bae76299/pkg/sql/sqlbase/kvfetcher.go#L227)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Yes we have ran tests for this and yes it is rather expensive. Sufficiently expensive in fact that Peter had added a separate boolean for this block of code (and related code) before, so it wouldn't run on the regular code path. Check the history of the code. @petermattis what do you think? Right now this code is alloc-heavy (especially the string appends). At the least use a bytes.Buffer.
</blockquote></details>

ah, I though we were arguing about the logging itself, not the code around it. `bytes.Buffer` is a good idea. Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19008)*
<!-- Sent from Reviewable.io -->

knz commented:
Reviewed 5 of 6 files at r6.
Review status: 7 of 8 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/sql/sqlbase/kvfetcher.go, line 227 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/19008#-KvZF0ROAeMSnNp1agq7:-Kvli7000G-3sp2ws02p:b-biair1) ([raw file](https://github.com/cockroachdb/cockroach/blob/eca98ea2a80df677baa30f90cd5076d8bae76299/pkg/sql/sqlbase/kvfetcher.go#L227)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

ah, I though we were arguing about the logging itself, not the code around it. `bytes.Buffer` is a good idea. Done.
</blockquote></details>

`buf.writeString(span.String())`  is still one allocation too many. Make a `Format` method on `span` and give it `buf` as argument.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19008)*
<!-- Sent from Reviewable.io -->

knz commented:
Also, same request as I made to Radu in #18508: please run a micro-benchmark (E.g. exactly the one Radu used) before and after this PR, and report the results. 1) I want to see the exact effects on allocations. @jordanlewis will be interested too. 2) it will give you (like everyone in the team is getting their chance too soon) experience with running these micro-benchmarks and attaching the results to the commit message as a matter of course.

RaduBerinde commented:
LGTM

jordanlewis commented:
LGTM

andreimatei commented:
Check out the new `ExpensiveLogEnabled()`.

---

Review status: 2 of 10 files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/sql/sqlbase/kvfetcher.go, line 227 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/19008#-KvZF0ROAeMSnNp1agq7:-Kvmz_X4NIstJjgQxfa2:b-dvhxh2) ([raw file](https://github.com/cockroachdb/cockroach/blob/eca98ea2a80df677baa30f90cd5076d8bae76299/pkg/sql/sqlbase/kvfetcher.go#L227)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

`buf.writeString(span.String())`  is still one allocation too many. Make a `Format` method on `span` and give it `buf` as argument.
</blockquote></details>

I'll give you satisfaction - the pretty printing of spans does cost a few alloc (assuming that's a meaningful "cost" :P).
I've gated this behind a new `ExpensiveLogEnabled()` - check it out.
I've added a link to bench results in the PR description.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19008#-:-Kvn-0e7Si_a1EODb4vE:b-7bic23)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 2 of 10 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/sql/sqlbase/kvfetcher.go, line 227 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/19008#-Kvn0VpjARpwX6gC0ImJ:-Kvn0VpjARpwX6gC0ImK:bbqt25g) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2d7d3ee3281e2c0101ebd2aea9d1ea629a74ebe/pkg/sql/sqlbase/kvfetcher.go#L227)):*
> ```Go
> 	}
> 
> 	if log.ExpensiveLogEnabled(ctx, 2) {
> ```

No! We want these spans do be printed in SHOW KV TRACE in all cases.

What about you restore the `traceKV` conditional for this one!?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19008)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 2 of 10 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/sql/sqlbase/kvfetcher.go, line 227 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/19008#-Kvn0VpjARpwX6gC0ImJ:-Kvn0uHwVvMt7CxcW0Wu:b-7mqqoc) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2d7d3ee3281e2c0101ebd2aea9d1ea629a74ebe/pkg/sql/sqlbase/kvfetcher.go#L227)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

No! We want these spans do be printed in SHOW KV TRACE in all cases.

What about you restore the `traceKV` conditional for this one!?
</blockquote></details>

They are printed with `SHOW [KV] TRACE`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19008)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 2 of 10 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/sql/sqlbase/kvfetcher.go, line 227 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/19008#-Kvn0VpjARpwX6gC0ImJ:-Kvn1CkrjWt_h7OrSBfG:b-bvasau) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2d7d3ee3281e2c0101ebd2aea9d1ea629a74ebe/pkg/sql/sqlbase/kvfetcher.go#L227)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

They are printed with `SHOW [KV] TRACE`.
</blockquote></details>

The conditional to establish this is complicated. Do you understand what SpanInContext does? Do you understand what VDepth does?  It's not much cheaper than hitting the heap allocator!
Testing a simple boolean is the  way to go.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19008)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 2 of 10 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/sql/sqlbase/kvfetcher.go, line 227 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/19008#-Kvn0VpjARpwX6gC0ImJ:-Kvn1whgOBIMBovV30lo:bvk7vif) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2d7d3ee3281e2c0101ebd2aea9d1ea629a74ebe/pkg/sql/sqlbase/kvfetcher.go#L227)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

The conditional to establish this is complicated. Do you understand what SpanInContext does? Do you understand what VDepth does?  It's not much cheaper than hitting the heap allocator!
Testing a simple boolean is the  way to go.
</blockquote></details>

Testing *which* boolean? The "kvTrace" one is the wrong one to trace.
I've given you the benchmarks and they show nothing. I'm telling you that we log a million things per KV request; the condition to evaluate whether this message should be constructed is not more expensive than any of the other a million `log.VEvent()` calls. What else can I do for you?
P.S. What do we do to not have this conversation for every `VEvent()` & friends call?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19008)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 2 of 10 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[pkg/util/log/log.go, line 192 at r9](https://reviewable.io:443/reviews/cockroachdb/cockroach/19008#-Kvo636bDM9sNvHWpT6m:-Kvo636bDM9sNvHWpT6n:b-7qa4mn) ([raw file](https://github.com/cockroachdb/cockroach/blob/ecc293e0ff7774b40f72046bb7e67f0f33ddc898/pkg/util/log/log.go#L192)):*
> ```Go
> // attached to ctx. In particular, if some OpenTracing collection is enabled
> // (e.g. LightStep), that, by itself, does NOT cause the expensive messages to
> // be enabled. SHOW TRACE FOR <stmt> and friends, on the other hand, does cause
> ```

[nit] do

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19008#-:-Kvo6waJDxrBMrLJbWI0:bnfp4nl)*
<!-- Sent from Reviewable.io -->

knz commented:
On 06-10-17 21:01, Andrei Matei wrote:
> I'm telling you that we log a million things per KV request; the
> condition to evaluate whether this message should be constructed is not
> more expensive than any of the other a million |log.VEvent()| calls.

The fact we have too much overhead on the hot path is undeniable. This
should be reduced, not increased.

> What else can I do for you?

A before/after benchmark comparison on an index join of 1000 rows, for
example, counting both the time overhead and the allocations.

-- 
Raphael 'kena' Poss

andreimatei committed:
sql: add an index-join benchmark

andreimatei committed:
sql: make traceNode message conditional

traceNode (used for SHOW TRACE FOR <stmt>) logs a message ("output row:
...") for every row produced by the inner stmt. These messages make
traces very long when many rows are produced, and also don't match the
behavior of SHOW TRACE FOR SESSION.
This patch makes it output that only for `SHOW _KV_ TRACE FOR <stmt>`.
Not that this is a particularly correct solution, but I think it's an
improvement.

andreimatei commented:
Review status: 2 of 11 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[pkg/sql/sqlbase/kvfetcher.go, line 227 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/19008#-Kvn0VpjARpwX6gC0ImJ:-Kw1HkwrAibfagSmXbmM:b7brcjz) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2d7d3ee3281e2c0101ebd2aea9d1ea629a74ebe/pkg/sql/sqlbase/kvfetcher.go#L227)):*
[Pasting here from your email response. Please be aware that email responds are posted at top comments, so they break conversations.]

The fact we have too much overhead on the hot path is undeniable. This should be reduced, not increased.

> What else can I do for you?

A before/after benchmark comparison on an index join of 1000 rows, for example, counting both the time overhead and the allocations.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19008)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Added a benchmark and another commit changing a log message in traceNode.

---

Review status: 2 of 11 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[pkg/sql/sqlbase/kvfetcher.go, line 227 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/19008#-Kvn0VpjARpwX6gC0ImJ:-Kw1IPFRbVDYrCbxhNm4:b-v35ae4) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2d7d3ee3281e2c0101ebd2aea9d1ea629a74ebe/pkg/sql/sqlbase/kvfetcher.go#L227)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

[Pasting here from your email response. Please be aware that email responds are posted at top comments, so they break conversations.]

The fact we have too much overhead on the hot path is undeniable. This should be reduced, not increased.

> What else can I do for you?

A before/after benchmark comparison on an index join of 1000 rows, for example, counting both the time overhead and the allocations.
</blockquote></details>

I added a commit with the benchmark. No difference.
```
[andrei:~/work/src/github.com/cockroachdb/cockroach] sql-kv-tracing+ ± benchstat /tmp/{before,after}
name         old time/op  new time/op  delta
IndexJoin-8  8.54ms ± 4%  8.54ms ± 1%   ~     (p=0.605 n=9+9)


[andrei:~/work/src/github.com/cockroachdb/cockroach] sql-kv-tracing+ ± benchstat /tmp/{before-mem,after-mem}
name         old time/op    new time/op    delta
IndexJoin-8    8.64ms ± 2%    8.59ms ± 6%    ~     (p=0.243 n=9+10)

name         old alloc/op   new alloc/op   delta
IndexJoin-8    3.62MB ± 0%    3.62MB ± 0%  -0.03%  (p=0.000 n=9+9)

name         old allocs/op  new allocs/op  delta
IndexJoin-8     11.6k ± 0%     11.6k ± 0%    ~     (p=0.314 n=10+10)
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19008#-:-Kw1IQgv0u7WrG0F4AVL:bdczo3b)*
<!-- Sent from Reviewable.io -->

knz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> modulo extending the commit message with the bench results

---

Reviewed 4 of 8 files at r10, 1 of 1 files at r11, 6 of 8 files at r12, 2 of 2 files at r13.
Review status: 11 of 12 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/sql/sqlbase/kvfetcher.go, line 227 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/19008#-Kvn0VpjARpwX6gC0ImJ:-Kw1moo7VgH3NdrkxMoQ:bnemfv1) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2d7d3ee3281e2c0101ebd2aea9d1ea629a74ebe/pkg/sql/sqlbase/kvfetcher.go#L227)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I added a commit with the benchmark. No difference.
```
[andrei:~/work/src/github.com/cockroachdb/cockroach] sql-kv-tracing+ ± benchstat /tmp/{before,after}
name         old time/op  new time/op  delta
IndexJoin-8  8.54ms ± 4%  8.54ms ± 1%   ~     (p=0.605 n=9+9)


[andrei:~/work/src/github.com/cockroachdb/cockroach] sql-kv-tracing+ ± benchstat /tmp/{before-mem,after-mem}
name         old time/op    new time/op    delta
IndexJoin-8    8.64ms ± 2%    8.59ms ± 6%    ~     (p=0.243 n=9+10)

name         old alloc/op   new alloc/op   delta
IndexJoin-8    3.62MB ± 0%    3.62MB ± 0%  -0.03%  (p=0.000 n=9+9)

name         old allocs/op  new allocs/op  delta
IndexJoin-8     11.6k ± 0%     11.6k ± 0%    ~     (p=0.314 n=10+10)
```
</blockquote></details>

Brilliant, thanks. Please copy this into the commit message.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19008#-:-Kw1n3LUCBgyHgG8rn45:b-lnevib)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 2 of 12 files reviewed at latest revision, 6 unresolved discussions.

---

*[pkg/sql/sqlbase/kvfetcher.go, line 227 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/19008#-Kvn0VpjARpwX6gC0ImJ:-Kw1sLsP2rxadhN-j6Rz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2d7d3ee3281e2c0101ebd2aea9d1ea629a74ebe/pkg/sql/sqlbase/kvfetcher.go#L227)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Brilliant, thanks. Please copy this into the commit message.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19008)*
<!-- Sent from Reviewable.io -->
